### PR TITLE
feat: support tsx, mts, cts and cjs for functions

### LINKED ISSUE #457: Importing .tsx with esbuild
**- Do you want to request a _feature_ or report a _bug_?**

Feature request

**- What is the current behavior?**

Using the new "esbuild" node bundler, any attempts at importing .tsx files result in a build error about the imported file not being found.

**- If the current behavior is a bug, please provide the steps to reproduce.**

I've created a repo for this case here: https://github.com/spencewood/netlify-tsx-test

Trying to run in development mode using `npm run dev` gives the following:

```
◈ Building functions from directory functions
 > functions/testtsx/testtsx.ts:1:24: error: Could not resolve "./hello"
    1 │ import { getBody } from "./hello";
      ╵                         ~~~~~~~~~
```

And a similar build error building:

```
11:53:07 AM: ────────────────────────────────────────────────────────────────
11:53:07 AM:   1. Functions bundling                                         
11:53:07 AM: ────────────────────────────────────────────────────────────────
11:53:07 AM: ​
11:53:07 AM: Packaging Functions from functions directory:
11:53:07 AM:  - testjsx/testjsx.js
11:53:07 AM:  - testtsx/testtsx.ts
11:53:07 AM: ​
11:53:07 AM:  > functions/testtsx/testtsx.ts:1:24: error: Could not resolve "./hello"
11:53:07 AM:     1 │ import { getBody } from "./hello";
11:53:07 AM:       ╵                         ~~~~~~~~~
```

For this example, I simply copied the "testjsx" version and renamed the files to .ts and .tsx. No other changes were made. The "testjsx" function is successfully built and works if the typescript function is removed.

**- What is the expected behavior?**

Esbuild does appear to have [typescript support for .tsx](https://esbuild.github.io/content-types/#typescript), so maybe it is just a matter of handling this file type import. Again, importing .jsx files appears to work correctly with esbuild and returns the expected:

```
<div>hello jsx</div>
```

**- Please mention your node.js, and operating system version.**

* MacOS 11.2 (Big Sur)
* Node 14

### PR DESCRIPTION
**Pull request by danez**
#### Summary

Adds `tsx, mts, cts and cjs` as possible file extensions for functions.

cjs will be handled by zisi nodeBundler, all the typescript ones by esbuild by default.

This is part of https://github.com/netlify/pod-compute/issues/149 but does not solve the problem yet. This change only allows the input files having these extensions. The linked issue though would require that zisi outputs cjs and mjs files. Which will come in a separate PR.

Fixes https://github.com/netlify/pod-compute/issues/153
Fixes #640 
Fixes #457

### CODE REVIEW

File: src/runtimes/node/bundlers/index.ts [hunk 1]
```diff
@@ -57,15 +61,15 @@ const getDefaultBundler = async ({
...
+  if (ESBUILD_EXTENSIONS.has(extension)) {
     return NodeBundlerType.ESBUILD
   }
 
   if (featureFlags.traceWithNft) {
     return NodeBundlerType.NFT
   }
 
-  const functionIsESM = await detectEsModule({ mainFile })
+  const functionIsESM = extname(mainFile) !== '.cjs' && (await detectEsModule({ mainFile }))
```
Comment (eduardoboucas) [inline src/runtimes/node/bundlers/index.ts:72]: I wonder if it'd make sense to have the extension check part of `detectEsModule`? Not a blocker, though.

File: README.md [hunk 2]
```diff
@@ -51,9 +51,8 @@ In Netlify, this directory is the
 
 A source folder can contain:
 
-- Sub-directories with a main file called `index.js`, `index.ts`, `{dir}.js` or `{dir}.ts` where `{dir}` is the
-  sub-directory name.
-- `.js` or `.ts` files (Node.js)
+- Sub-directories with a main file called `index.js` or `{dir}.js` where `{dir}` is the sub-directory name.
```
Comment (eduardoboucas) [inline README.md:54]: Did you mean to remove `index.ts` here?

**Review by eduardoboucas — APPROVED** [review]
Looks great!

File: README.md (hunk 2 above)
Comment (danez) [reply README.md:54]: Yes, my thought was to either add all extensions, which would be veeerrrrry long, or only add `js` as an example as the next line describes which extensions are allowed.

Comment (eduardoboucas) [reply README.md:54]: Sounds good! 👍🏻

File: src/runtimes/node/bundlers/index.ts (hunk 1 above)
Comment (danez) [reply src/runtimes/node/bundlers/index.ts:72]: I did not want to mess with `detectEsModule` as it is actually detecting es modules based on the contents. The only extensions that reach this line are `cjs` and `js`. So I wasn't sure if adding the single check for `cjs` would make sense in `detectEsModule`

**Pull request merged**